autorally_core/src/autorally_chassis/AutoRallyChassis.cpp

```diff
@@ -0,0 +1,583 @@
+{
+  switch(msgType[0])
+  {
+    //wheel speeds data as comma separated doubles, units in m/s
+    case 'w':
+    {
+      std::vector<std::string> data;
+      boost::split(data, msg, boost::is_any_of(","));
+      if(data.size() == 4)
+      {
+        autorally_msgs::wheelSpeedsPtr wheelSpeeds(new autorally_msgs::wheelSpeeds);
+        try
+        {
+          wheelSpeeds->lfSpeed = std::stod(data[0]);
+          wheelSpeeds->rfSpeed = std::stod(data[1]);
+          wheelSpeeds->lbSpeed = std::stod(data[2]);
+          wheelSpeeds->rbSpeed = std::stod(data[3]);
+          if(wheelSpeedsPub_ && !ros::isShuttingDown())
+          {
+            wheelSpeeds->header.stamp = ros::Time::now();
+            wheelSpeedsPub_.publish(wheelSpeeds);
+          }
+          serialPort_.tick("wheelSpeeds data");
+        } catch (std::exception& e )
+        {
+          serialPort_.diag_warn("Converting wheel speeds data failed");
+        }
+      } else
+      {
+        serialPort_.diag_warn("Processing wheel speeds data failed");
+      }
+      break;
+    }
+    
+    //Actuator controls from RC input, as comma separated us pulse width, currentl frontBrake is not controlled by RC
+    case 'r':
+    {
+      std::vector<std::string> data;
+      boost::split(data, msg, boost::is_any_of(","));
+      if(data.size() == 4)
+      {
+        autorally_msgs::chassisCommandPtr chassisCommand(new autorally_msgs::chassisCommand);
+        try
+        {
+          //std::cout << std::stoi(data[0]) << " " << std::stoi(data[1]) << std::endl;
+          chassisCommand->steering = actuatorUsToCmd(std::stoi(data[0]), "steering");
+          chassisCommand->throttle = actuatorUsToCmd(std::stoi(data[1]), "throttle");
+          chassisCommand->frontBrake = -5.0;
+          chassisCommand->sender = "RC";
+          //this line is in here for compatibility with the old servoInterface
+          chassisCommand->header.frame_id = "RC";
+          
+          rcMutex_.lock();
+          mostRecentRcSteering_ = chassisCommand->steering;
+          mostRecentRcThrottle_ = chassisCommand->throttle;
+          mostRecentRcFrontBrake_ = chassisCommand->frontBrake;
+          rcMutex_.unlock();
+          
+          if(chassisCommandPub_ && !ros::isShuttingDown())
+          {
+            chassisCommand->header.stamp = ros::Time::now();
+            chassisCommandPub_.publish(chassisCommand);
+          }
+
+          chassisStateMutex_.lock();
+          if(std::stoi(data[2]) > 1500)
+          {
+            autonomousEnabled_ = true;
+          } else
+          {
+            autonomousEnabled_ = false;
+          }
+          throttleRelayEnabled_ = std::stoi(data[3]); //this value can be 0 or 1
+          chassisStateMutex_.unlock();
+
+          serialPort_.tick("RC data");
+        } catch(std::exception& e)
+        {
+          serialPort_.diag_warn("Converting wheel speeds data failed");
```

Comment:
Except, I think this method locks the port, we would need to find a workaround if we keep it locked in the above function.